aten/src/ATen/native/Onehot.cpp

```diff
@@ -0,0 +1,35 @@
+#include <ATen/ATen.h>
+
+namespace at { namespace native {
+
+Tensor one_hot(const Tensor &self, int64_t num_classes) {
+    AT_ASSERTM(self.dtype() == kLong, "one_hot is only applicable to index tensor.");
+    auto shape = self.sizes().vec();
+
+    // empty tensor could be converted to one hot representation,
+    // but shape inference is not possible.
+    if (self.numel() == 0) {
+        if (num_classes <= 0) {
+            AT_ERROR("Can not infer total number of classes from empty tensor.");
+        } else {
+            shape.push_back(num_classes);
+            return at::empty(shape, self.options());
+        }
+    }
+
+    // non-empty tensor
+    AT_ASSERTM(self.min().item().toLong() >= 0, "Class values must be non-negative.");
+    if (num_classes == -1) {
+        num_classes = self.max().item().toLong() + 1;
+    } else {
+        AT_ASSERTM(num_classes > self.max().item().toLong(), "Class values must be smaller than num_classes.");
```

Comment:
Use `AT_ERROR` instead. It might be better for user usability if it were something like
```
AT_ERROR(num_classes > self.max().item().toLong(), "Class values (", self.max().item().toLong(), ") must be smaller than num_classes (", num_classes, ")");
```